src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -237,7 +237,15 @@ private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
             var workspaceRegistration = (WorkspaceRegistration)sender;
             uint docCookie;
-            Contract.ThrowIfFalse(_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie));
+
+            // external workspace that listens to events from IVSRunningDocumentTable4 (just like MiscellaneousFilesWorkspace does) and registers the text buffer for opened document prior to Miscellaneous workspace getting notified of the docCookie - the prior contract assumes that MiscellaneousFilesWorkspace is always the first one to get notified
+            if (!_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie))
+            {
+                // We haven't even started tracking the document corresponding to this registration - likely because some external workspace registered the document's buffer prior to MiscellaneousWorkspace getting notified of it.
```

Comment:
I wouldn't think this is possible -- that's why I had the ThrowIfFalse. In particular this comment seems off:

> likely because some external workspace registered the document's buffer prior to MiscellaneousWorkspace getting notified of it.

We only should be getting events for registrations for specific text buffers we've subscribed to; the comment implies a misunderstanding of when this event is raised. This would imply that:

1) we got the event before we added the registration -> doc cookie map on line 226 which should be impossible given this should all be single-threaded. Note that code runs in response to us seeing a doc cookie, so if somebody else sees and registers before us, that's totally fine. That's what line 228 is written to deal with.
2) we got this event after we removed the registration in the various places where we do that, which might mean we have other bugs too since I thought that was synchronous.

The comment implies the first, but if that's the case then the bug is elsewhere and this is not the correct fix. If it's the latter, we might have more problems to fix. Can you clarify @mavasani?